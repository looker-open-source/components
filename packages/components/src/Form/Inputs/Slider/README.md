# Slider

`Slider` is used for selecting a single value. If you need to specify a range of values, use the [RangeSlider component](?path=/docs/docs-rangeslider--docs) instead.

## Default Slider

The `<Slider />` component renders a styled range slider on the page, with no accompanying label.

```tsx
function Basic(props: SliderProps) {
  return <Slider {...props} />;
}
```

## Value, Min, Max

The `<Slider />` component accepts a `value` property, allowing the user to set the initial value at which the slider is initialized.
It also accepts `min` and `max` values.

```tsx
function MinMaxValue(props: SliderProps) {
  const { min = 10, max = 20, value = 15, ...restProps } = props;

  return <Slider min={min} max={max} value={value} {...restProps} />;
}
```

Min and max default to the following values:

- **min**: 0
- **max**: 10

The slider component will disregard any value prop that goes outside of the defined min or max. The following example illustrates default settings when `max=10` but `value=5000`

```tsx
function InvalidValue(props: SliderProps) {
  const { min = 0, max = 10, value = 5000, ...restProps } = props;

  return <Slider min={min} max={max} value={value} {...restProps} />;
}
```

## Step

Just like the standard HTML range input, you change the granularity of slider options by using the `step` prop.

```tsx
function Step(props: SliderProps) {
  const { max = 10000, step = 2500, ...restProps } = props;

  return <Slider max={max} step={step} {...restProps} />;
}
```

## Accessibility

Slider accepts the standard aria-labelledby to help screen readers associate the range input with the relevant form label.

```tsx
function LabeledSlider(props: SliderProps) {
  return (
    <>
      <Label id="slider-label">Slider:</Label>
      <Slider aria-labelledby="slider-label" {...props} />
    </>
  );
}
```

## Disabled

Use the disable property to gray out the Input, making it a fixed value.

```tsx
function Disabled(props: SliderProps) {
  const { disabled = true, ...restProps } = props;

  return <Slider disabled={disabled} {...restProps} />;
}
```

## Controlled component

If you'd like to read and control the input value externally, `Slider` accepts an onChange event callback.

```tsx
function Controlled(props: SliderProps) {
  const { value: valueProp = 5, onChange: _onChange, ...restProps } = props;

  const [sliderValue, setSliderValue] = useState(valueProp);
  const onChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSliderValue(event.target.value);
  };

  return (
    <Space>
      <Slider onChange={onChange} value={sliderValue} {...restProps} />
      <Paragraph>
        <strong>Value:</strong> {sliderValue}
      </Paragraph>
    </Space>
  );
}
```
