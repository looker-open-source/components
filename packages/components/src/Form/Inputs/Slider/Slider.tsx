/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { forwardRef, Ref, SyntheticEvent, useState } from 'react'
import isFunction from 'lodash/isFunction'
import styled, { css } from 'styled-components'
import {
  reset,
  space,
  SpaceProps,
  typography,
  TypographyProps,
} from '@looker/design-tokens'
import { WidthProps, width } from 'styled-system'

import { InputProps } from '../InputProps'

export interface SliderProps
  extends SpaceProps,
    WidthProps,
    Omit<InputProps, 'type'>,
    TypographyProps {
  'aria-labelledby'?: string
  max?: number
  min?: number
  step?: number
  value?: number
}

const SliderInternal = forwardRef(
  (
    {
      min = 0,
      max = 10,
      value = 0,
      step,
      onChange,

      name,
      id,
      className,
      disabled,
      ...restProps
    }: SliderProps,
    ref: Ref<HTMLInputElement>
  ) => {
    const [isFocused, setIsFocused] = useState(false)
    const [internalValue, setInternalValue] = useState(value)

    if (min > max) {
      // Props don't make sense. ABORT!!
      // eslint-disable-next-line no-console
      console.warn(
        `Unable to render <Slider /> because the 'min' prop was set greater than 'max' value. MIN: ${min}, MAX: ${max}`
      )
      return null
    }

    const boundSliderValue = (v: number) => Math.min(Math.max(v, min), max) // enforce that value stays between min and max

    const displayValue = isFunction(onChange)
      ? boundSliderValue(value)
      : boundSliderValue(internalValue)

    const fillPercent = Math.round(((displayValue - min) / (max - min)) * 100)

    const handleFocus = () => {
      setIsFocused(true)
    }

    const handleUnfocus = () => {
      setIsFocused(false)
    }

    const internalChangeHandler = (event: SyntheticEvent<HTMLInputElement>) => {
      const evtValue = (event.target as HTMLInputElement).value
      setInternalValue(parseInt(evtValue))
    }

    const handleChange = isFunction(onChange) ? onChange : internalChangeHandler
    return (
      <div className={className} data-testid="container">
        <SliderValueWrapper>
          <SliderValue
            disabled={disabled}
            isFocused={isFocused}
            offsetPercent={fillPercent}
          >
            {displayValue}
          </SliderValue>
        </SliderValueWrapper>
        <SliderTrack>
          <SliderFill offsetPercent={fillPercent} disabled={disabled} />
          <SliderThumb
            isFocused={isFocused}
            offsetPercent={fillPercent}
            disabled={disabled}
          />
        </SliderTrack>
        <SliderInput
          disabled={disabled}
          id={id}
          isFocused={isFocused}
          max={max}
          min={min}
          name={name}
          onChange={handleChange}
          step={step}
          offsetPercent={fillPercent}
          value={displayValue}
          aria-labelledby={restProps['aria-labelledby']}
          data-testid="slider-input"
          ref={ref}
          onBlur={handleUnfocus}
          onFocus={handleFocus}
        />
      </div>
    )
  }
)

interface SliderInputProps {
  isFocused?: boolean
  disabled?: boolean
  offsetPercent: number
}

const sliderThumbCss = css`
  height: 16px;
  visibility: hidden;
  width: 16px;
`

const SliderThumb = styled.div<SliderInputProps>`
  border-radius: 100%;
  left: ${({ offsetPercent = 0 }) => `${offsetPercent}%`};
  position: absolute;
  top: -6px;
  transform: translateX(-50%);
  transition: transform 250ms, box-shadow 250ms;

  ${({ theme: { colors }, isFocused, disabled }) => css`
    background: ${colors.field};
    border: 3px solid ${colors.key};
    height: 16px;
    width: 16px;
    ${isFocused && 'border-width: 5px;'}
    ${disabled && `border-color: ${colors.neutral};`}
  `}
`

const SliderInput = styled.input.attrs(() => ({ type: 'range' }))<
  SliderInputProps
>`
  background: transparent;
  display: block;
  height: 22px;
  left: 8px;
  margin-left: 0;
  margin-right: 0;
  position: relative;
  -webkit-appearance: none; /* stylelint-disable-line */
  width: calc(100% - 16px);

  &::-webkit-slider-thumb {
    -webkit-appearance: none; /* stylelint-disable-line */
    ${sliderThumbCss}
  }

  &::-moz-range-thumb {
    ${sliderThumbCss}
  }

  &::-ms-thumb {
    ${sliderThumbCss}
  }

  &::-ms-track {
    background: transparent;
    border-color: transparent;
    color: transparent;
    cursor: pointer;
    width: 100%;
  }

  &::-moz-focus-outer {
    border: none;
  }

  &:focus {
    outline: none;
  }
`

const SliderTrack = styled.div`
  background: ${({ theme }) => theme.colors.ui2};
  border-radius: ${({ theme }) => theme.radii.small};
  height: 4px;
  left: 16px;
  margin-top: -2px;
  position: absolute;
  top: 50%;
  width: calc(100% - 32px);
`

interface ControlProps {
  offsetPercent: number
  disabled?: boolean
}

const SliderFill = styled.div<ControlProps>`
  background: ${({ theme: { colors }, disabled }) =>
    disabled ? colors.neutral : colors.key};
  border-radius: ${({ theme }) => theme.radii.small};
  height: 100%;
  width: ${({ offsetPercent }) => offsetPercent}%;
`

interface SliderValueProps extends SliderInputProps {
  disabled?: boolean
  isFocused: boolean
  offsetPercent: number
}

const SliderValue = styled.div<SliderValueProps>`
  background: ${({ theme, isFocused }) =>
    isFocused ? theme.colors.keyAccent : theme.colors.keyText};
  border-radius: 1rem;
  color: ${({ theme: { colors }, disabled }) =>
    disabled ? colors.neutral : colors.key};
  left: ${({ offsetPercent }) => offsetPercent}%;
  padding: 0 0.5rem;
  position: absolute;
  text-align: center;
  transform: translateX(-50%) translateY(-0.9rem);
  user-select: none;
`

const SliderValueWrapper = styled.div`
  margin: 0 auto;
  position: relative;
  width: calc(100% - 30px);
`

export const Slider = styled(SliderInternal)<SliderProps>`
  ${reset}
  ${space}
  ${width}
  ${typography}
  position: relative;
`

SliderInternal.displayName = 'Slider'
Slider.defaultProps = {
  fontSize: 'small',
  lineHeight: 'xsmall',
  mt: 'medium',
  width: '100%',
}
