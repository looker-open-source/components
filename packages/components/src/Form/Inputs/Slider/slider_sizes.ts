/*

 MIT License

 Copyright (c) 2019 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  SizeLarge,
  SizeMedium,
  SizeSmall,
  FontSizes,
} from '@looker/design-tokens'
import { variant } from 'styled-system'

export type SliderSizes = SizeSmall | SizeMedium | SizeLarge

export interface SliderSizeProps {
  /**
   * Defines the size of the slider.
   * @default "medium"
   */
  size?: SliderSizes
}

interface SliderSizeMap {
  prop: 'size'
  variants: {
    [key in SliderSizes]: {
      knobSize: number
      trackHeight: number
      fontSize: FontSizes
      valueSpacing: string
    }
  }
}

/* eslint-disable sort-keys */
const sliderSizeMap: SliderSizeMap = {
  prop: 'size',
  variants: {
    small: {
      knobSize: 12,
      trackHeight: 3,
      fontSize: 'xxsmall',
      valueSpacing: '-0.75rem',
    },
    medium: {
      knobSize: 16,
      trackHeight: 4,
      fontSize: 'xsmall',
      valueSpacing: '-0.85rem',
    },
    large: {
      knobSize: 18,
      trackHeight: 6,
      fontSize: 'small',
      valueSpacing: '-1rem',
    },
  },
}
/* eslint-enable sort-keys */

export const sliderSize = variant(sliderSizeMap)
