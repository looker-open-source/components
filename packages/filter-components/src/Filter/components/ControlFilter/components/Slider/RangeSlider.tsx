/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { FC } from 'react'
import React from 'react'
import styled from 'styled-components'
import isNumber from 'lodash/isNumber'
import { RangeSlider as RangeSliderComponent } from '@looker/components'
import type { RangeSliderProps } from './types'

export const RangeSlider: FC<RangeSliderProps> = ({
  value,
  options,
  onChange,
}) => {
  const bounds = { min: 0, max: 100, ...(options || {}) }
  const handleOnChange = (v: number[]) => {
    onChange({ min: v[0], max: v[1] })
  }

  return (
    <RangeSliderWrapper>
      <RangeSliderComponent
        max={bounds.max}
        min={bounds.min}
        value={[
          isNumber(value.min) ? value.min : bounds.min,
          isNumber(value.max) ? value.max : bounds.max,
        ]}
        onChange={handleOnChange}
      />
    </RangeSliderWrapper>
  )
}

const RangeSliderWrapper = styled.div`
  font-size: ${({ theme }) => theme.fontSizes.small};
  min-width: 150px;
  padding: 0 ${({ theme }) => theme.space.small};
  width: 100%;
`
