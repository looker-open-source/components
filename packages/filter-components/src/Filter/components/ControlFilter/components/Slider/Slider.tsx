/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { FC, SyntheticEvent } from 'react'
import React from 'react'
import styled from 'styled-components'
import { Slider as SliderComponent } from '@looker/components'
import type { SliderProps } from './types'

export const Slider: FC<SliderProps> = ({ onChange, value, options }) => {
  const optionsWithDefaults = { min: 0, max: 100, ...options }

  const handleChange = (e: SyntheticEvent<HTMLInputElement>) => {
    const { value } = e.currentTarget
    onChange(parseInt(value, 10))
  }

  return (
    <SliderWrapper>
      <SliderComponent
        min={optionsWithDefaults.min}
        max={optionsWithDefaults.max}
        value={value}
        onChange={handleChange}
      />
    </SliderWrapper>
  )
}

const SliderWrapper = styled.div`
  font-size: ${({ theme }) => theme.fontSizes.small};
  width: 100%;
`
