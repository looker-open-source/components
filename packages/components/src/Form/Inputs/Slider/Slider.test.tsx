/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { fireEvent } from '@testing-library/react'
import {
  // mountWithTheme,
  // assertSnapshot,
  renderWithTheme,
} from '@looker/components-test-utils'

import { Slider } from './Slider'

test('Enforces that input value can never go above `max` prop value', () => {
  const props = {
    max: 1,
    value: 2,
  }
  const { getByTestId } = renderWithTheme(<Slider {...props} />)
  const input = getByTestId('slider-input') as HTMLInputElement
  expect(parseInt(input.value)).toEqual(props.max)
})

test('Enforces that input value can never go below `min` prop value', () => {
  const props = {
    min: 5,
    value: 1,
  }

  const { getByTestId } = renderWithTheme(<Slider {...props} />)
  const input = getByTestId('slider-input') as HTMLInputElement
  expect(parseInt(input.value)).toEqual(props.min)
})

test('Slider with name and id', () => {
  const props = {
    id: 'Slip',
    name: 'Slide',
  }
  const { getByTestId } = renderWithTheme(<Slider {...props} />)
  const input = getByTestId('slider-input')
  expect(input).toHaveAttribute('id', props.id)
  expect(input).toHaveAttribute('name', props.name)
})

test('Slider input can be disabled', () => {
  const { getByTestId } = renderWithTheme(<Slider disabled />)
  const input = getByTestId('slider-input') as HTMLInputElement
  expect(input.disabled).toEqual(true)
})

test('Accessibility: Slider with aria-labeledby and <label>', () => {
  const { getByLabelText, getByTestId } = renderWithTheme(
    <>
      <label id="some-id">Slider Label</label>
      <Slider aria-labelledby="some-id" />
    </>
  )

  expect(getByLabelText('Slider Label')).toEqual(getByTestId('slider-input'))
})

test('Slider passes change event to optional onChange handler', () => {
  const handleChange = jest.fn()
  const { getByTestId } = renderWithTheme(
    <Slider min={0} max={10} onChange={handleChange} />
  )

  expect(handleChange).toHaveBeenCalledTimes(0)

  fireEvent.change(getByTestId('slider-input'), {
    target: { value: 10 },
  })

  expect(handleChange).toHaveBeenCalledTimes(1)
})

test('renders focus styles on keypress', () => {
  const { getByTestId } = renderWithTheme(<Slider />)

  fireEvent.keyUp(getByTestId('container'), { key: 'Tab', keyCode: 9 })

  expect(getByTestId('slider-input')).toMatchSnapshot()
})
