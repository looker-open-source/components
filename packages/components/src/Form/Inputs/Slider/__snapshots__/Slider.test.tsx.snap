// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders focus styles on keypress 1`] = `
.c0 {
  background: transparent;
  display: block;
  height: 22px;
  left: 8px;
  margin-left: 0;
  margin-right: 0;
  position: relative;
  -webkit-appearance: none;
  width: calc(100% - 16px);
}

.c0::-webkit-slider-thumb {
  -webkit-appearance: none;
  height: 16px;
  visibility: hidden;
  width: 16px;
}

.c0::-moz-range-thumb {
  height: 16px;
  visibility: hidden;
  width: 16px;
}

.c0::-ms-thumb {
  height: 16px;
  visibility: hidden;
  width: 16px;
}

.c0::-ms-track {
  background: transparent;
  border-color: transparent;
  color: transparent;
  cursor: pointer;
  width: 100%;
}

.c0::-moz-focus-outer {
  border: none;
}

.c0:focus {
  outline: none;
}

<input
  class="c0"
  data-testid="slider-input"
  max="10"
  min="0"
  type="range"
  value="0"
/>
`;
